### PR TITLE
ddl: add data structure `SubJob` for multi-schema change

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/14766

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/cancel_test.go [hunk 1]
```diff
@@ -25,19 +25,49 @@ import (
...
+func testMatchCancelState(t *testing.T, job *model.Job, cancelState interface{}, sql string) bool {
+	switch v := cancelState.(type) {
+	case model.SchemaState:
+		if job.Type == model.ActionMultiSchemaChange {
+			msg := fmt.Sprintf("unexpected multi-schema change(sql: %s, cancel state: %s)", sql, v)
+			assert.Failf(t, msg, "use []model.SchemaState as cancel states instead")
+			return false
+		}
+		return job.SchemaState == v
+	case []model.SchemaState: // For multi-schema change sub-jobs.
```
Comment (tangenta) [inline ddl/cancel_test.go:53]:
```suggestion
	case subStates: // For multi-schema change sub-jobs.
```

File: util/dbterror/ddl_terror.go [hunk 2]
```diff
@@ -34,6 +34,11 @@ var (
 	ErrCancelledDDLJob = ClassDDL.NewStd(mysql.ErrCancelledDDLJob)
 	// ErrRunMultiSchemaChanges means we run multi schema changes.
 	ErrRunMultiSchemaChanges = ClassDDL.NewStdErr(mysql.ErrUnsupportedDDLOperation, parser_mysql.Message(fmt.Sprintf(mysql.MySQLErrName[mysql.ErrUnsupportedDDLOperation].Raw, "multi schema change"), nil))
+	// ErrOperateSameColumn means we change the same columns multiple times in a DDL.
+	ErrOperateSameColumn = ClassDDL.NewStdErr(mysql.ErrUnsupportedDDLOperation, parser_mysql.Message(fmt.Sprintf(mysql.MySQLErrName[mysql.ErrUnsupportedDDLOperation].Raw, "operate same column '%s'"), nil))
+	// ErrOperateSameIndex means we change the same indexes multiple times in a DDL.
+	ErrOperateSameIndex = ClassDDL.NewStdErr(mysql.ErrUnsupportedDDLOperation, parser_mysql.Message(fmt.Sprintf(mysql.MySQLErrName[mysql.ErrUnsupportedDDLOperation].Raw, "operate same index '%s'"), nil))
+	// ErrWaitReorgTimeout means we wait for reorganization
```
Comment (tangenta) [inline util/dbterror/ddl_terror.go:41]:
```suggestion
```

File: table/tables/tables.go [hunk 3]
```diff
@@ -1456,9 +1456,6 @@ func GetColDefaultValue(ctx sessionctx.Context, col *table.Column, defaultVals [
 	if col.GetOriginDefaultValue() == nil && mysql.HasNotNullFlag(col.GetFlag()) {
 		return colVal, errors.New("Miss column")
 	}
-	if col.State != model.StatePublic {
```
Comment (wjhuang2016) [inline table/tables/tables.go:1459]: Why remove it? Is it safe?

Comment (tangenta) [reply table/tables/tables.go:1459]: Consider the case:

```sql
create table t (a tinyint);
insert into t set a = 10 ;
alter table t add column b int not null, change column a c char(5) first;
```

In this ALTER TABLE statement, we add a **NOT NULL** column first, and then do the lossy column type change. In the phase of backfilling, the schema state of column `b` is still "write-reorg". If we keep this code, a null value is assigned to the column `b`, which violates the not-null constraint. Also see https://github.com/tangenta/tidb/issues/94.

I don't see if there are any benefits to get a null value for the middle-state column... so let it be the real default value.

No tests are broken by this change.

**Review by wjhuang2016 — APPROVED** [review]

**Review by xiongjiwei — APPROVED** [review]

**Pull request merged**